File: typescript/test/techniques.test.ts
import { describe, expect, test } from 'vitest';
import { hiddenSingle, nakedSingle } from '../src/techniques';
import {
  boardEligibleForHiddenSingle,
  boardEligibleForNakedSingle,
  boardNotEligibleForSingle,
  buildBoard,
} from './fixtures';

describe('techniques', () => {
  describe('nakedSingle', () => {
    test('find naked single', () => {
      const board = buildBoard(boardEligibleForNakedSingle);
      const row = 0;
      const col = 0;
      // Cell is not yet solved
      expect(board.cells[col][row].getValue()).toBeUndefined();

      board.initializeCandidates();
      const boardWithProgress = nakedSingle(board);
      expect(boardWithProgress.board.cells[row][col].getValue()).toBeDefined();
      expect(boardWithProgress.progress).toEqual(true);
    });

    test('find nothing', () => {
      const board = buildBoard(boardEligibleForHiddenSingle);
      const row = 0;
      const col = 0;
      // Cell is not yet solved
      expect(board.cells[row][col].getValue()).toBeUndefined();

      board.initializeCandidates();
      const boardWithProgress = nakedSingle(board);
      expect(
        boardWithProgress.board.cells[row][col].getValue()
      ).toBeUndefined();
      expect(boardWithProgress.progress).toEqual(false);
    });
  });

  describe('hiddenSingle', () => {
    test('find hidden single', () => {
      const board = buildBoard(boardEligibleForHiddenSingle);
      const row = 0;
      const col = 0;
      // Cell is not yet solved
      expect(board.cells[col][row].getValue()).toBeUndefined();

      board.initializeCandidates();
      const boardWithProgress = hiddenSingle(board);
      expect(boardWithProgress.board.cells[row][col].getValue()).toBeDefined();
      expect(boardWithProgress.progress).toEqual(true);
    });

    test('find nothing', () => {
      const board = buildBoard(boardNotEligibleForSingle);
      const row = 0;
      const col = 3;
      // Cell is not yet solved
      expect(board.cells[row][col].getValue()).toBeUndefined();

      board.initializeCandidates();
      const boardWithProgress = hiddenSingle(board);
      expect(boardWithProgress.board.cells[3][0].getValue()).toBeUndefined();
      expect(boardWithProgress.progress).toEqual(false);
    });
  });
});

File: kotlin/src/test/kotlin/TechniquesTest.kt
import dev.patientallison.sudoku.data.Coordinates
import dev.patientallison.sudoku.nakedSingle
import org.junit.jupiter.api.Nested
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

class TechniquesTest {
    @Nested
    inner class NakedSingle {
        @Test
        fun `Find naked single`() {
            val board = buildBoard(nakedSingleBoard)
            val coordinates = Coordinates(0, 0)
            // Cell is not yet solved
            assertNull(board.getCellFromCoordinates(coordinates).getValue())

            board.initializeCandidates()
            val boardWithProgress = nakedSingle(board)
            assertNotNull(boardWithProgress.board.getCellFromCoordinates(coordinates).getValue())
            assertEquals(true, boardWithProgress.progress)
        }

        @Test
        fun `Find nothing`() {
            val board = buildBoard(hiddenSingleBoard)
            val coordinates = Coordinates(0, 0)
            // Cell is not yet solved
            assertNull(board.getCellFromCoordinates(coordinates).getValue())

            board.initializeCandidates()
            val boardWithProgress = nakedSingle(board)
            // Cell is still not solved, no progress was made
            assertNull(boardWithProgress.board.getCellFromCoordinates(coordinates).getValue())
            assertEquals(false, boardWithProgress.progress)
        }
    }
}
